### PR TITLE
[FrameworkBundle] integrate the Cache component

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17537 |
| License | MIT |
| Doc PR | - |

Last commit is the diff with #17868.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/cache_pools.xml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+
+        <service id="cache.adapter.default" alias="cache.adapter.filesystem" />
+
+        <service id="cache.adapter.apcu" class="Symfony\Component\Cache\Adapter\ApcuAdapter" abstract="true">
+            <tag name="cache.pool" namespace-arg-index="0" default-lifetime-arg-index="1" />
+            <argument /> <!-- namespace -->
+            <argument /> <!-- default lifetime -->
```
Comment (Koc) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/cache_pools.xml]: In other services `namespace` is second argument but here is first. Maybe better change constructor signature for better consistency?

Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/cache_pools.xml]: But default lifetime is always second...

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -547,4 +548,39 @@ private function addPropertyInfoSection(ArrayNodeDefinition $rootNode)
...
+                                        ->defaultValue('default')
+                                    ->end()
+                                    ->booleanNode('public')->defaultFalse()->end()
+                                    ->integerNode('default_lifetime')->defaultNull()->end()
+                                    ->scalarNode('provider_service')->defaultNull()->end()
+                                    ->scalarNode('directory')->defaultNull()->end()
+                                ->end()
+                                ->validate()
+                                    ->ifTrue(function ($v) {
+                                        return isset($v['provider_service']) xor in_array($v['type'], array('psr6', 'doctrine'));
```
Comment (Nicofuma) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: According to the error message it should be 
`return isset($v['provider_service']) xor !in_array($v['type'], array('psr6', 'doctrine'));`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            }
+            foreach ($tags[0] as $argName => $argValue) {
+                if (isset($tags[0][$argName.'_arg_index']) && 0 <= $tags[0][$argName.'_arg_index']) {
+                    $pool->replaceArgument($tags[0][$argName.'_arg_index'], $argValue);
+                }
+            }
+        }
+    }
+
+    private function getNamespace($id)
```
Comment (trousers) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php]: Codacy Issue found: [Avoid variables with short names like $id. Configured minimum length is 3.](https://www.codacy.com/app/git_9/symfony/file/2331802099/issues/source?bid=3207821&fileBranchId=3207834#l56)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class CachePoolPass implements CompilerPassInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
```
Comment (trousers) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php]: Codacy Issue found: [The method process() has a Cyclomatic Complexity of 11. The configured cyclomatic complexity threshold is 10.](https://www.codacy.com/app/git_9/symfony/file/2331802099/issues/source?bid=3207821&fileBranchId=3207834#l27)

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php [hunk 5]
```diff
@@ -636,4 +650,47 @@ private function assertVersionStrategy(ContainerBuilder $container, Reference $r
             $this->assertEquals($format, $versionStrategy->getArgument(1));
         }
     }
+
+    private function assertCachePoolServiceDefinitionIsCreated(ContainerBuilder $container, $name, $type, $defaultLifetime, $namespaceArgumentIndex = null)
```
Comment (trousers) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: Codacy Issue found: [Avoid excessively long variable names like $namespaceArgumentIndex. Keep variable name length under 20.](https://www.codacy.com/app/git_9/symfony/file/2331802253/issues/source?bid=3207821&fileBranchId=3207834#l654)

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php [hunk 6]
```diff
@@ -636,4 +650,47 @@ private function assertVersionStrategy(ContainerBuilder $container, Reference $r
...
+    {
+        $id = 'cache.pool.'.$name;
```
Comment (trousers) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: Codacy Issue found: [Avoid variables with short names like $id. Configured minimum length is 3.](https://www.codacy.com/app/git_9/symfony/file/2331802253/issues/source?bid=3207821&fileBranchId=3207834#l656)

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php (hunk 5 above)
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: @trousers are you a bot? If so, I fear we get spammed. We use fabbot for automatic code reviews and I really doubt we want an other bot we can't configure here.
About the comments, I'd really like to get comments on the actual behavior of the code before getting more on the length of variables.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/cache_pools.xml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        </service>
+
+        <service id="cache.adapter.doctrine" class="Symfony\Component\Cache\Adapter\DoctrineAdapter" abstract="true">
+            <tag name="cache.pool" service-provider-arg-index="0" default-lifetime-arg-index="1" namespace-arg-index="2" />
+            <argument /> <!-- doctrine provider service -->
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/cache_pools.xml]: Doctrine

**Comment by xabbuh** [comment]
I am not convinced that people will need to create services starting with `cache.adapter.` to register their own adapters. This will prevent us from being able to add new adapters in the core in future releases as we would otherwise risk to break BC. I think we should, for now, restrict the configuration to the built-in adapters. We can then think of a clean way to let users register their own adapters for 3.2 (probably with another yet to be introduced DI tag).

Besides that I am :+1: on the current state of this feature.

**Comment by xabbuh** [comment]
ping @symfony/deciders

**Comment by nicolas-grekas** [comment]
@xabbuh second review please :)

**Comment by xabbuh** [comment]
Not sure it's the best idea from the DX perspective when I have to pass the service name instead of a simple key, but I think we can live with that. :)

So to me this is ready to be merged.

ping @symfony/deciders

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/cache_pools.xml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        <service id="cache.adapter.default" alias="cache.adapter.filesystem" />
+
+        <service id="cache.adapter.apcu" class="Symfony\Component\Cache\Adapter\ApcuAdapter" abstract="true">
+            <tag name="cache.pool" namespace-arg-index="0" default-lifetime-arg-index="1" />
+            <argument /> <!-- namespace -->
+            <argument /> <!-- default lifetime -->
+        </service>
+
+        <service id="cache.adapter.doctrine" class="Symfony\Component\Cache\Adapter\DoctrineAdapter" abstract="true">
+            <tag name="cache.pool" service-provider-arg-index="0" default-lifetime-arg-index="1" namespace-arg-index="2" />
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/cache_pools.xml]: Do we really need the `default-lifetime-arg-index` and `namespace-arg-index` parameters? They are so ugly :sob: and uncommon in Symfony codebase.

**Comment by fabpot** [comment]
I'm really not a big fan of the `arg-index` thing.

**Comment by nicolas-grekas** [comment]
Please suggest any alternatives...

**Comment by HeahDude** [comment]
Since it should be documented maybe the "arg" part could be dropped.

**Comment by javiereguiluz** [comment]
@nicolas-grekas the alternative would be to replace this handy argument resolver:

```php
foreach ($tags[0] as $argName => $argValue) {
    if (isset($tags[0][$argName.'_arg_index']) && 0 <= $tags[0][$argName.'_arg_index']) {
        $pool->replaceArgument($tags[0][$argName.'_arg_index'], $argValue);
    }
}
```

And replace it with a `switch ... case` to treat the arguments of each cache type differently.

**Comment by HeahDude** [comment]
@javiereguiluz it's a very nice workaround to use the better name, but would `default-lifetime="0"` not be misleading ?

**Comment by xabbuh** [comment]
The drawback of that is that we wouldn't be able to support every kind of custom cache adapter a user might want to create, but we would stick to the built-in adapters instead.

**Comment by nicolas-grekas** [comment]
This PR introduces two concepts:
- cache pools: each cache pool is one single logical set of cache items of the same kind and in the same scope (e.g. one pool for caching SQL queries, another pool for caching html fragments, another one for caching parsed annotations, etc.). The kind of storage (apcu, redis, etc.) is a technical requirement but not a semantic requirement for the definition of a cache pool.
- cache adapters: a cache adapter is well, a php class that implements a particular storage.

This PR provides a way to manipulate both concepts separately:
- app authors need to define cache pools. This is easily done by using the semantic configuration so that one can configure e.g. cache.pools.foobar and that's it. By default, the concrete storage implementation will be the one implemented by the `cache.adapter.default` service, which is cache.adapter.filesystem by default. But the main idea it that when one creates cache.pools.foobar, it's a one-liner.

Then, devs can work on optimizing the cache storages. There are two ways to do so:
- the first is to redefine the cache.adapter.default service/alias. Again, it's a one-liner in the service definition file to redefine it to e.g. cache.adapter.apcu instead of cache.adapter.filesystem. This way, all cache pools that have no explicit adapter configured switch to apcu storage instead of filesystem.
- the second way is per cache pool, by explicitly configuring a cache adapter for each of them. Using semantic configuration, this is done with the "type" option. A "type" here, is the name of another cache pool, whose adapter configuration is reused (by service decoration).

On the service definitions side, a cache pool is any service that has the cache.pool tag.
A cache adapter is any service that has the cache.pool tag AND is abstract.
Concrete cache pools are usually defined as decorated cache adapters or pools definitions.
Thus, we need a way to map configuration parameters from concrete definitions to decorated definitions, and this is what is cache.pool is used for: its attributes allow one to say: I want a cache pool that decorates the cache.adapter.apcu service with a default lifetime of 60s and a namespace of foobar. Then the *-arg-index attributes, defined at the abstract cache.adapter.apcu service definition level, allow mapping these namespace+default-lifetime to the correct constructor argument of the apcu service class.

In short: defining *-arg-index attributes won't be part of the daily work to do on a Symfony app. The list of cache storage adapter implementations is short. This PR provides a few, a bundle can provide more.
What people will do in there app is reuse these. They won't have to manipulate *arg-index attributes. Still, *-arg-index attributes provides extensibility, which is an important target to me.

**Comment by nicolas-grekas** [comment]
Btw, `*-arg-index` could be made useless if the DI component had a "parameterized abstract service definition decoration" feature.

**Comment by nicolas-grekas** [comment]
I replaced `*arg-index` wiring by a conventional constructor arguments ordering.
Better DYT?

**Comment by javiereguiluz** [comment]
@nicolas-grekas much better! Thanks.

**Comment by nicolas-grekas** [comment]
Just added and wired a Psr6CacheClearer so that the e.g. cache:clear command clears all cache.pool tagged services.

File: src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php [hunk 9]
```diff
@@ -57,9 +57,9 @@ public function __construct()
         );
 
         $this->removingPasses = array(
+            new ReplaceAliasByActualDefinitionPass(),
             new RemovePrivateAliasesPass(),
             new RemoveAbstractDefinitionsPass(),
-            new ReplaceAliasByActualDefinitionPass(),
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: This looks like a bug fix to me, is it? Aliases should be resolved before removing private and abstract definitions, isn't it?

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php [hunk 10]
```diff
@@ -96,11 +97,12 @@ private function resolveArguments(ContainerBuilder $container, array $arguments,
      */
     private function resolveDefinition(ContainerBuilder $container, DefinitionDecorator $definition)
     {
-        if (!$container->hasDefinition($parent = $definition->getParent())) {
+        try {
+            $parentDef = $container->findDefinition($parent = $definition->getParent());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php]: This is a new feature of theDI component: being able to decorate an aliased definition.

File: src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php [hunk 11]
```diff
@@ -58,8 +58,8 @@ public function __construct()
 
         $this->removingPasses = array(
             new RemovePrivateAliasesPass(),
-            new RemoveAbstractDefinitionsPass(),
             new ReplaceAliasByActualDefinitionPass(),
+            new RemoveAbstractDefinitionsPass(),
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: This looks like a bug fix to me: aliases should be resolved before removing abstract definitions, isn't it?

**Comment by nicolas-grekas** [comment]
All green, ping @symfony/deciders 
I also added functional tests, with an example configuring a default Redis adapter in https://github.com/symfony/symfony/pull/18371/files#diff-3ed79b3403a07c5e56afea9b4023eae2

**Comment by fabpot** [comment]
:+1: BUT I'd like to get feedback from the community about the way things are configured; it's powerful but not easy. Documentation is going to help, but let see what users think about the current state. We still have about 2 months to tweak the configuration and we still have the option to remove this before 3.1 if it proves to be difficult to find a better way.

File: src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php (hunk 11 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php]: Fixed on 2.3 in #18447

**Comment by nicolas-grekas** [comment]
Since this is not yet merged, I added one more commit:
- add a `cache.pool.default` service so that one could use caching without configuring anything to begin with (ping @javiereguiluz )
- throw LogicException when one want to configure a pool but the Cache component is not loaded

**Comment by javiereguiluz** [comment]
@nicolas-grekas to check if this is optimized for the expected default behavior, please tell me what should a developer do by hand to achieve this:

1. I have a server with APCu enabled.
2. I execute `symfony new my_project` to create a 3.1-based project
3. I want to boost the performance of my app. I want to cache anything that is cacheable: serializer, entities metadata, validator, property path, forms, etc. (If possible, third-party bundles too)

And now, what should that same developer do to achieve this:

1. I've installed Redis in that same server and now I want to keep caching everything, but using Redis instead of APCu.

**Comment by nicolas-grekas** [comment]
@javiereguiluz in fact our expectations are too low here: Cache integration should not be limited to pre-computed information about the app itself. We should also ease caching of cluster wide items. APCu is not suited for them, because APCu doesn't allow a shared cache pool amongst e.g. several fronts.
This means we should allow grouping cache pools by cache items scope: APCu is perfect for reflection-related info. Redis is a good fit for cluster wide items.
My recommendation would be to configure at least two main abstract pools. They would both default to a filesystem cache adapter, but one could then easily change the configuration of each of those abstract pools, effectively reconfiguring all concrete pools that rely on them.
The goal of this PR is *not* to do this configuration, but to provide the basis to work on it in a later PR (during the feature freeze 3.1).
As an example, the current state of this PR is to provide a single `cache.pool.default`, that is designed to store only app-related items.
Symfony-related things should go in other pools that we'll add later.
As you can read in the attached functional tests, changing all adapters at once is as easy as redefining the `cache.adapter.default` service (in one test, the default filesystem adapter is changed for a redis one).

**Comment by nicolas-grekas** [comment]
The Cache component is now a mandatory dep of the framework bundle, and a `cache.adapter.system` is created that auto-selects apcu when available at compile time and is the adapter to use for any reflection-related cache pools.

**Comment by mvrhov** [comment]
@nicolas-grekas if the apcu is automatic this will probably bring problems when deploys are made. e.g the deploy would require php-fpm or apache restart.

**Comment by nicolas-grekas** [comment]
@mvrhov if you're referring to restarting for the purpose of cache clearing, this is already handled with a cache clearer, so that clearing the cache does not require any reboot but is automatic when the cache:clear command is called.
Or did you have something else in mind?

**Comment by mvrhov** [comment]
I haven't looked at the driver implementation, but I'm pretty sure that in the case of apcu is not. Because you cannot clear the apcu cache from cli

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+            'provider',
+            'namespace',
+            'default_lifetime',
+        );
+        foreach ($container->findTaggedServiceIds('cache.pool') as $id => $tags) {
+            $adapter = $pool = $container->getDefinition($id);
+            if ($pool->isAbstract()) {
+                continue;
+            }
+            if (!isset($tags[0]['namespace'])) {
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php]: Could it make sense to apply several ``cache.pool`` tags on the service ? If yes, we need to handle all tags, not only the first one.
If not, we may want to warn the user that subsequent tags are ignored to ease debugging

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php [hunk 13]
```diff
@@ -96,11 +97,12 @@ private function resolveArguments(ContainerBuilder $container, array $arguments,
...
+        } catch (ServiceNotFoundException $e) {
             throw new RuntimeException(sprintf('The parent definition "%s" defined for definition "%s" does not exist.', $parent, $this->currentId));
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php]: please link the exception as a previous exception, as it might give more details in case of a chain of aliases (telling which alias of the chain failed to resolve)

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php [hunk 14]
```diff
@@ -96,11 +97,12 @@ private function resolveArguments(ContainerBuilder $container, array $arguments,
      */
     private function resolveDefinition(ContainerBuilder $container, DefinitionDecorator $definition)
     {
-        if (!$container->hasDefinition($parent = $definition->getParent())) {
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php]: another solution is to use ``$container->has()`` to validate the direct parent, keeping the existing error message as is, and then simply using ``findDefinition`` below (and letting it throw its exception if an alias cannot be resolved, as the message will be clear in this case)

File: src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+use Psr\Cache\CacheItemPoolInterface;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class Psr6CacheClearer implements CacheClearerInterface
+{
+    private $pools = array();
+
+    public function addPool(CacheItemPoolInterface $pool)
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php]: I suggest making the cache clearer immutable, by injecting an array of pools in the constructor instead. What do you think ?

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePools/redis_config.yml [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+        class: Redis
+        calls:
+            - [connect, [127.0.0.1]]
+
+    cache.adapter.default:
+        abstract: true
+        parent: cache.adapter.redis
+        tags:
+            - name: cache.pool
+              provider: cache.adapter.redis.connection
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePools/redis_config.yml]: please make it namespaced, to avoid clearing the whole local redis when running tests

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php (hunk 10 above)
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php]: btw, this new feature must have tests covering it and a changelog entry (and it would be great to submit it as a dedicated PR)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php (hunk 12 above)
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php]: I doesn't make sense. We can add a warning, but unless I'm wrong no other compiler passes do that...

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePools/redis_config.yml (hunk 16 above)
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePools/redis_config.yml]: it is already namespaced by the compiler pass

File: src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php (hunk 15 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php]: Since each cache pools can define its own clearer, that would complicate building these arrays I think. That's why I did it this way.

**Comment by nicolas-grekas** [comment]
Online & offline comments addressed (hello from Symfony Live Paris).
The config now defines two default cache adapters: `cache.adapter.shared` & `cache.adapter.local`,
and two cache pools for the user/dev side of the app only: `cache.pool.shared` & `cache.pool.local`.
We'll have to create symfony-related pools in other PRs.
I also reverted the auto-magic selecting apcu when avail: filesystem is now always the default, don't forget to change the config when you need something else. (e.g. cache.adapter.redis for cache.adapter.shared and cache.adapter.apcu for cache.adapter.local).

**Comment by mvrhov** [comment]
@nicolas-grekas: even if you removed the apcu being default, the problem with it still persists. IMO this requires a fat warning message in the docs. Because probably most people are not aware of that problem.

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php (hunk 14 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php]: done like that

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveDefinitionTemplatesPassTest.php [hunk 17]
```diff
@@ -309,6 +309,20 @@ public function testProcessMergeAutowiringTypes()
         $this->assertEquals(array('Foo', 'Bar'), $def->getAutowiringTypes());
     }
 
+    public function testProcessResolvesAliases()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveDefinitionTemplatesPassTest.php]: @stof test case added

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.
